frappe/public/js/frappe/form/controls/date.js

```diff
@@ -75,12 +75,13 @@ frappe.ui.form.ControlDate = frappe.ui.form.ControlData.extend({
 		if(!this.frm) return;
 		// show datepicker above or below the input
 		// based on scroll position
-		var window_height = $(window).height();
-		var window_scroll_top = $(window).scrollTop();
-		var el_offset_top = this.$input.offset().top + 280;
-		var position = 'top left';
-		if(window_height + window_scroll_top >= el_offset_top) {
-			position = 'bottom left';
+		const scroll_limit = $(window).scrollTop() + $(window).height();
+		const picker_bottom = this.$input.offset().top + this.$input.outerHeight() + 12 + this.datepicker.$datepicker.outerHeight();
+
+		var position = 'bottom left';
+		//if (picker_top > scroll_limit + window_scroll) {
+		if (picker_bottom >= scroll_limit) {
+			position = 'top left';
```

Comment:
What does these changes do? Before / After screenshots?

Reply:
The old system preferred to open 'down' when possible, which tends to cover the next box (since people work from top to bottom). It also relied on a fixed magic number of 270.
![OLD_prefer_down](https://user-images.githubusercontent.com/4930097/68622034-137b3280-04c9-11ea-8092-3aee1a1fbf1f.png)
It also tended to get it wrong and overlap at the bottom of the window when it should have opened upwards.
![OLD_overlapping_date](https://user-images.githubusercontent.com/4930097/68621936-d57e0e80-04c8-11ea-9bf0-2e155ff1b2b9.png)
![OLD_overlapping_time](https://user-images.githubusercontent.com/4930097/68622044-1aa24080-04c9-11ea-8907-0b25a98f323f.png)
![OLD_overlapping_datetime](https://user-images.githubusercontent.com/4930097/68622048-1e35c780-04c9-11ea-9226-08a9cc7bf3b9.png)

Reply:
New method is (almost) entirely positioned according to screen elements (except fix for broken time-picker which is off by 42px when opening up) and correctly opens down instead of up when there is not space below the header.
Not quite enough space to open 'up' (so opens down, which overlaps bottom of the window slightly as the picker is too long to fit either way):
![NEW_go_down_just](https://user-images.githubusercontent.com/4930097/68622261-a74cfe80-04c9-11ea-8e07-3e8f8747351a.png)
Slightly scrolled up window with just enough space to open upwards:
![NEW_go_up_just](https://user-images.githubusercontent.com/4930097/68622288-b6cc4780-04c9-11ea-9bb1-4d6ee7d65764.png)